### PR TITLE
Feat AMM

### PR DESCRIPTION
**Pull request by peterslany**
Merge AMM branch back into master.

### CODE REVIEW

File: src/utils/loans.ts
Comment (bvotteler) [inline src/utils/loans.ts:21]: Naming between this method and `calculateTotalBorrowedBtcChange` (ends with "Change") above seems to be inconsistent. I might be missing something, though.

File: test/unit/parachain/loans.test.ts
Comment (bvotteler) [inline test/unit/parachain/loans.test.ts:300]: Not sure about this, but do we need `await`s here to avoid promises being still evaluated after this test suite is marked as complete?

File: test/unit/utils/rewards.test.ts
Comment (bvotteler) [inline test/unit/utils/rewards.test.ts:25]: Similar here, I suspect we might need an `await` here, so the throw doesn't happen outside of the `describe` block if it occurs. (Not 100% sure tho)

**Review by bvotteler — COMMENTED** [review]
Looking good for the most part. Left comments on the points needing clarification.

Expecting quite a few changes around the tests in particular once master is merged back in.

File: src/parachain/loans.ts
Comment (gregdhill) [inline src/parachain/loans.ts:352]: `amount` is unused

File: src/parachain/amm.ts
Comment (gregdhill) [inline src/parachain/amm.ts:869]: Won't block on this, but just wondering if we could type "Percentage" to make it explicit what value should be passed?

File: src/utils/currency.ts
Comment (gregdhill) [inline src/utils/currency.ts:151]: This is a nice change, should be easier to close this issue now: https://github.com/interlay/interbtc-api/issues/599

**Review by gregdhill — COMMENTED** [review]
We should be able to remove most of the path finding code once this is merged: https://github.com/interlay/interbtc/pull/944

File: scripts/docker-setup.sh
Comment (peterslany) [reply scripts/docker-setup.sh]: Yes, this script was already removed form master, so removing from this PR as well :+1:

File: package.json
Comment (peterslany) [reply package.json:3]: I used `1.21.0`, I would increase major version only if we do so on parachain side too to stay consistent. what do you think?

File: src/parachain/amm.ts
Comment (peterslany) [reply src/parachain/amm.ts:836]: Left over comment, thanks for finding!

File: src/parachain/amm/liquidity-pool/stable.ts
Comment (peterslany) [reply src/parachain/amm/liquidity-pool/stable.ts]: This constant is used as a precision point when approximating pool invariant variables using Newton's method, I moved it to `constants` file and added more informative comment and name.

File: src/utils/loans.ts
Comment (peterslany) [reply src/utils/loans.ts]: You are right, changed naming to make it consistent :)

File: test/unit/parachain/loans.test.ts
Comment (peterslany) [reply test/unit/parachain/loans.test.ts:301]: No, in this case there is no need to await, because `getLendingStats` does not return promise.

File: test/unit/utils/rewards.test.ts
Comment (peterslany) [reply test/unit/utils/rewards.test.ts]: This was removed from master, removing from PR.

File: package.json
Comment (bvotteler) [reply package.json]: The way I look at minor version bumps for the lib is: If we have a breaking change in an existing API we definitely bump at least the minor version.

Given these changes are quite big, with whole new groups of APIs available, I think a minor version bump is justified.

And we're not aligned with clients/parachain versions anyway, so I feel we can justify a bit more wriggle room in our choices :)

File: package.json
Comment (bvotteler) [inline package.json:4]: Paging @gregdhill , I seem to recall you removed this ("and kBTC") in your other PR?

File: src/parachain/rewards.ts
Comment (bvotteler) [inline src/parachain/rewards.ts:64]: TIL you can share a nonce across `queryMulti` queries. Nice!

File: src/utils/constants.ts
Comment (bvotteler) [inline src/utils/constants.ts:26]: Wasn't that `.pow(18)` before? Or is this a different value?

File: src/utils/currency.ts
Comment (bvotteler) [inline src/utils/currency.ts:266]: Do we need an `.isSome` check before unwrapping, or is it guaranteed to be set?

File: src/utils/rewards.ts
Comment (bvotteler) [inline src/utils/rewards.ts:4]: Not sure where this is used, but would it make sense to get an rpc for that at some stage?

We've pushed these types of calcs for other scenarios to rpcs - granted some of them were more complex, so it made sense to keep the logic in the parachain code. Just something to consider

**Review by bvotteler — COMMENTED** [review]
LGTM, just a few smaller things to clarify.

File: test/utils/helpers.ts
Comment (gregdhill) [inline test/utils/helpers.ts:212]: Already declared as a getter on `InterBtcApi`

File: package.json
Comment (peterslany) [reply package.json]: Makes sense, I'll tag it as `2.0.0` then :+1:

File: src/utils/constants.ts
Comment (peterslany) [reply src/utils/constants.ts:26]: It was 18, but I reduced it to 9 to make the computation faster - the approximations are still precise enough for the swap amounts computations.

File: src/utils/currency.ts
Comment (peterslany) [reply src/utils/currency.ts:266]: Yes, this is always set if the lendToken exists. If not the function will throw - which is correct behaviour for invalid `lendTokenId` provided.

File: test/utils/helpers.ts
Comment (peterslany) [reply test/utils/helpers.ts]: Right, I'll reuse that one then :+1:

File: src/utils/rewards.ts
Comment (bvotteler) [reply src/utils/rewards.ts:4]: > Yes, I think adding rpc that would provide calculation with more accurate average block time makes sense!

cc @gregdhill

**Review by bvotteler — APPROVED** [review]
LGTM

**Pull request merged**
